### PR TITLE
Implement CopyObjectRequest.

### PR DESCRIPTION
**Pull request by coryan**
This is part of the work for #812. It implements the request object
for `Objects: copy`.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1085)

### CODE REVIEW

**Comment by coryan** [comment]
Ping.

File: google/cloud/storage/internal/object_requests.h [hunk 1]
```diff
@@ -125,6 +125,43 @@ class InsertObjectStreamingRequest
...
+class CopyObjectRequest
+    : public GenericRequest<
+          CopyObjectRequest, DestinationPredefinedAcl, IfGenerationMatch,
+          IfGenerationNotMatch, IfMetagenerationMatch, IfMetagenerationNotMatch,
+          IfSourceGenerationMatch, IfSourceGenerationNotMatch,
+          IfSourceMetagenerationMatch, IfSourceMetagenerationNotMatch,
+          Projection, SourceGeneration, UserProject> {
+ public:
+  CopyObjectRequest() = default;
+  CopyObjectRequest(std::string destination_bucket,
```
Comment (houglum) [inline google/cloud/storage/internal/object_requests.h]: Any reason the destination args are placed before the source args here?  Semantically, I'd expect source args to come before dest args.

Comment (coryan) [reply google/cloud/storage/internal/object_requests.h]: No particular reason, I have seen both: `std::memcpy()` using destination first, `std::copy()` using destination second. Happy to change it if you think it is more natural in the reverse order.

Comment (coryan) [reply google/cloud/storage/internal/object_requests.h]: Hmmm, I just realized that @frankyn change for compose  (#1090) uses destination first too. Let's keep both the same?

Comment (frankyn) [reply google/cloud/storage/internal/object_requests.h]: No particular reason either. I don't mind reordering. I understand @houglum's point.

Comment (coryan) [reply google/cloud/storage/internal/object_requests.h]: I have reversed the order of the arguments

**Comment by coryan** [comment]
PTAL.

**Pull request merged**